cupyx/_texture.py

```diff
@@ -0,0 +1,170 @@
+import cupy
+
+from cupy import _core
+from cupy.cuda import texture
+from cupy.cuda import runtime
+
+
+_affine_transform_2d_array_kernel = _core.ElementwiseKernel(
+    'U texObj, raw float32 m, uint64 width', 'T transformed_image',
+    '''
+    float3 pixel = make_float3(
+        (float)(i / width) + .5f,
+        (float)(i % width) + .5f,
+        1.0f
+    );
+    float x = dot(pixel, make_float3(m[0],  m[1],  m[2]));
+    float y = dot(pixel, make_float3(m[3],  m[4],  m[5]));
+    transformed_image = tex2D<T>(texObj, y, x);
+    ''',
+    'affine_transformation_2d_array',
+    preamble='''
+    inline __host__ __device__ float dot(float3 a, float3 b)
+    {
+        return a.x * b.x + a.y * b.y + a.z * b.z;
+    }
+    ''')
+
+
+_affine_transform_3d_array_kernel = _core.ElementwiseKernel(
+    'U texObj, raw float32 m, uint64 height, uint64 width',
+    'T transformed_volume',
+    '''
+    float4 voxel = make_float4(
+        (float)(i / (width * height)) + .5f,
+        (float)((i % (width * height)) / width) + .5f,
+        (float)((i % (width * height)) % width) + .5f,
+        1.0f
+    );
+    float x = dot(voxel, make_float4(m[0],  m[1],  m[2],  m[3]));
+    float y = dot(voxel, make_float4(m[4],  m[5],  m[6],  m[7]));
+    float z = dot(voxel, make_float4(m[8],  m[9],  m[10], m[11]));
+    transformed_volume = tex3D<T>(texObj, z, y, x);
+    ''',
+    'affine_transformation_3d_array',
+    preamble='''
+    inline __host__ __device__ float dot(float4 a, float4 b)
+    {
+        return a.x * b.x + a.y * b.y + a.z * b.z + a.w * b.w;
+    }
+    ''')
+
+
+def _create_texture_object(data,
+                           address_mode: str,
+                           filter_mode: str,
+                           read_mode: str,
+                           border_color=0):
+
+    if cupy.issubdtype(data.dtype, cupy.unsignedinteger):
+        fmt_kind = runtime.cudaChannelFormatKindUnsigned
+    elif cupy.issubdtype(data.dtype, cupy.integer):
+        fmt_kind = runtime.cudaChannelFormatKindSigned
+    elif cupy.issubdtype(data.dtype, cupy.floating):
+        fmt_kind = runtime.cudaChannelFormatKindFloat
+    else:
+        raise ValueError(f'Unsupported data type {data.dtype}')
+
+    if address_mode == 'nearest':
+        address_mode = runtime.cudaAddressModeClamp
+    elif address_mode == 'constant':
+        address_mode = runtime.cudaAddressModeBorder
```

Comment:
I had to remove modes "wrap" and "mirror". If cudaTextureDesc::normalizedCoords is set to zero, cudaAddressModeWrap and cudaAddressModeMirror won't be supported and will be switched to cudaAddressModeClamp.

Reply:
I also renamed "border" and "clamp", to be consistent with scipy mode names.

Reply:
I was refreshing my memory on the texture memory. `normalizedCoords` is not easy to use in this context, right?

Reply:
It should be possible to use normalized coordinates and that in turn should allow using mirror and wrap texture modes. I will take a look in the next few days, maybe it is not too much effort. Thanks for the suggestion!